### PR TITLE
AutoML with holdout dataset and improved descriptions

### PR DESCRIPTION
**Pull request by unmonoqueteclea**
In order to avoid any kind of overoptimistic evaluations in AutoML, we are
creating a holdout dataset from the beginning, different from the
validation dataset used internally in Feature Selection stage.

So this new holdout dataset is not used until the final evaluation
step.

OptiML does not receive a test-dataset any more, so it will perform
cross-validation. It makes no sense using a specific validation
dataset in the OptiML, because this dataset won't be provided by the
user, it will be sampled from the provided train-datasets.

I also improved the descriptions of some of the scripts as propsed by Guillem (/cc @guillembigml)

### CODE REVIEW

**Comment by charleslparker** [comment]
@unmonoqueteclea - Thinking on this a bit more; I think it might still be a good idea to still hold out data for the OptiML that you do not use during feature selection.  This doesn't.have any impact on the final evaluation, but I think if you cross validate in the model on the same data you used for feature selection, you will end up with a bias in model selection towards models that can take advantage of that information leak (though it's not immediately clear to me how this might play out in the various types of models we offer).

This is closely related to the issue raised in the Hastie book.  As he says, all of the evaluations in the OptiML will be artificially higher if you cross validate in the OptiML.  That's easy to see.  What I'm wondering here is if the *relative quality* will also be different.  That is, will the evaluations in the OptiML be higher by different amounts in a way that will cause you to choose the wrong models for the fusion?

In my opinion, the odds of that happening are high enough where you should play it safe:  Split your training set 80/20, do feature selection on the 80% then run OpitML with the 80% as training data and 20% as validation.  Select the best models from the OptiML, fuse, and test on the provided test set.

**Comment by unmonoqueteclea** [comment]
Thanks, @charleslparker, I added a new holdout dataset for OptiML as you suggested.
Ready to be reviewed, @mmerce.

File: automl/automl-library/library.whizzml [hunk 1]
```diff
@@ -529,15 +536,13 @@
...
+(define (create-optiml-model train valid params shallow)
   (when (= (resource-type train) "dataset")
     (log-info "Creating Optiml, this  can last several minutes (or even hours)")
-    (let (max-time {"max_training_time" (if shallow 900 3600)}
-          params (merge params
-                        (if (= (resource-type validation) "dataset")
-                          (merge max-time {"test_dataset" validation})
-                          max-time)))
-      (create-optiml train params))))
+    (let (model-params (merge {"max_training_time" (if shallow 900 3600)
```
Comment (mmerce) [inline automl/automl-library/library.whizzml:542]: here do you want the user parameters to have precedence over these max_training_time and test_dataset?

File: automl/automl-script/metadata.json [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
... holdout and test datasets, the following tasks will\nbe automatically done:\n\n-  `Unsupervised Models Generation`: Creating the following\n  **unsupervised models**: `Cluster`, `Anomaly Detector`, `Association\n  Discovery` (with leverage and lift as search_strategy), `PCA` and\n**  `Topic Model` (for datasets that contain text fields).**\n\n - `Feature Generation`: Using the unsupervised models created\n  previously to append automatically generated new features to all the\n**  user given datasets.**\n\n - `Feature Selection`: Reducing, automatically, the number of fields\n  of the datasets using the **Recursive Feature Eliminination**\n  algorithm. This step is bypassed if the total number of fields is\n**  lower than 50 (or 30 if `shallow_search` is `True`)**\n\n - `Model Selection` Using OptiML to find the best models and using\n  the top 3 models to create a `Fusion` model to predict all the test\n  dataset instances. The script will return an evaluation of the final\n  model too.\n\n\nFind more info in [package readme](https://github.com/whizzml/examples/tree/master/automl) and [script readme](https://github.com/whizzml/examples/blob/master/automl/automl-script/readme.md)",
```
Comment (mmerce) [inline automl/automl-script/metadata.json]: It need**s** the AutoML...

File: automl/automl-script/metadata.json [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
-      "description": "Train dataset id (e.g. dataset/5d272205eba31d61920005cd)"
+      "description": "Train dataset id"
     },
     {
-      "name": "validation-dataset",
+      "name": "holdout-dataset",
       "type": "dataset-id",
       "default": "",
-      "description": "Validation dataset id (e.g. dataset/5d272205eba31d61920005cd)"
+      "description": "Hold-out dataset id, only to perform final evaluation"
```
Comment (mmerce) [inline automl/automl-script/metadata.json]: Hold-out dataset ID. Used to evaluate the final Fusion model

File: automl/automl-script/metadata.json [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
     },
     {
       "name": "test-dataset",
       "type": "dataset-id",
       "default": "",
-      "description": "Test dataset id (e.g. dataset/5d272205eba31d61920005cd)"
+      "description": "Test dataset id, for scoring"
```
Comment (mmerce) [inline automl/automl-script/metadata.json]: Test dataset ID. Inputs that will be scored using the final Fusion model

File: automl/automl-script/readme.md [hunk 5]
```diff
@@ -21,24 +21,28 @@ be automatically done:
...
   the top 3 models to create a `Fusion` model to predict all the test
-  dataset instances. If a validation dataset is given, this script
-  will return an evaluation of the final model too.
+  dataset instances. The script will return an evaluation of the final
+  model too.
 
 
 The **inputs** for the script are:
 
 * `train-dataset`: (dataset-id) Dataset id for the train dataset
```
Comment (mmerce) [inline automl/automl-script/readme.md]: dataset ID fro the train...

File: automl/automl-script/readme.md [hunk 6]
```diff
@@ -21,24 +21,28 @@ be automatically done:
...
-* `validation-dataset`: (dataset-id) Dataset id for the validation
-  dataset (e.g. dataset/5d272205ea31d61920005cd). If empty, some
-  random rows from the `train-dataset` will be set aside and used as
-  `validation-dataset`. The parameter `validation-rate` indicates the
-  portion of the `train-dataset` that is sampled in the
-  `validation-dataset`
+  provided, the script will expect an `automl-execution` and will use
+  its models as starting point. Useful when we only want to perform
+  scoring , using  a previous execution of AutoML.
+* `holdout-dataset`: (dataset-id) Dataset id for the holdout dataset
```
Comment (mmerce) [inline automl/automl-script/readme.md]: dataset ID

File: automl/automl-script/readme.md [hunk 7]
```diff
@@ -21,24 +21,28 @@ be automatically done:
...
+  (e.g. dataset/5d272205ea31d61920005cd). It will be used **only at
+  the end of the workflow** after the model has been trained, to
```
Comment (mmerce) [inline automl/automl-script/readme.md]: ... after the resulting fusion model has been created, to evaluate it.

File: automl/automl-script/readme.md [hunk 8]
```diff
@@ -21,24 +21,28 @@ be automatically done:
...
+  evaluate the final model. If empty, some random rows from the
+  `train-dataset` will be set aside and used as `holdout-dataset`. The
+  parameter `holdout-rate` indicates the portion of the
+  `train-dataset` that is sampled in the `holdout-dataset`
 * `test-dataset`: (dataset-id) Dataset id for the test dataset
```
Comment (mmerce) [inline automl/automl-script/readme.md]: Dataset ID...

**Comment by mmerce** [comment]
Let me see if I understand the present status:

- We allow the user to set 3 datasets: training, test and holdout
- We extend them with the features created from unsupervised learning models
- We split the training dataset to hold out a validation-dataset (used only in OptiML selection)
- We split the remaining training dataset to hold out a holdout if the user gives us a holdout rate (used only in the Fusion evaluation at the end)
- Now we have [training, validation, holdout, (and maybe) test].

Evaluations:
- In RFE we split again the training part to hold out a different validation dataset to do the feature selection.
- In OptiML we use the validation dataset held out in step 3 to do the type+parameters selection,
- We create a Fusion and an evaluation using the holdout-dataset
For this part, every time we run AutoML we get different splits, so non-deterministic behaviour, right? Which I assume is itended to be able to test consistent results over different executions.

Testing:
Iff there's a test-dataset:
- We create a batch-prediction for the test-dataset based on the previous Fusion

@unmonoqueteclea, please correct me if I'm wrong

File: automl/automl-script/readme.md [hunk 9]
```diff
@@ -21,24 +21,28 @@ be automatically done:
   lower than 50 (or 30 if `shallow_search` is `True`)
 -  `Model Selection` Using OptiML to find the best models and using
   the top 3 models to create a `Fusion` model to predict all the test
-  dataset instances. If a validation dataset is given, this script
-  will return an evaluation of the final model too.
+  dataset instances. The script will return an evaluation of the final
+  model too.
```
Comment (mmerce) [inline automl/automl-script/readme.md]: I would replace the last line of the previous paragraph and place a new item: 

- Fusion evaluation: Evaluating the Fusion using the data in the holdout dataset
- Fusion prediction: Scoring the input data provided in the test dataset (if set) using the final Fusion model.

File: automl/automl-script/readme.md [hunk 10]
```diff
@@ -95,14 +99,24 @@ overwritten by the corresponding input in `automl-execution` if this
 is given.
 
 **WARNING** All the fields that appear as `non-preferred` in train,
-validation or test datasets will be considered `non-preferred` in all
+holdout or test datasets will be considered `non-preferred` in all
 the resources created by AutoML. It doesn't matter in which dataset
-(train, validation or test) they are set as non-preferred fields.
+(train, holdout or test) they are set as non-preferred fields.
+
+**WARNING** In order to know your AutoML model performance, you should
```
Comment (mmerce) [inline automl/automl-script/readme.md]: The AutoML script creates partial models and evaluations both in the **RFE** and the **OptiML** selection processes. Those models and evaluations are using samples of the training data and independent validation datasets (extracted from the training data as totally complementary splits to avoid leakage between the Feature Selection and Model Selection steps). Therefore, only the final evaluation of the Fusion model on the holdout data can be used as an estimate for its performance.

File: automl/automl-script/readme.md [hunk 11]
```diff
@@ -111,23 +125,26 @@ The **outputs** for the script are:
 ## Usage
 There are two different ways of using this script:
 
-### From a train and a test dataset (and an optional validation dataset)
+### From a train and a test dataset (and an optional holdout dataset)
 In this case, the expected inputs for the script are the
 `train-dataset` and the `test-dataset`, with no `autml-execution`
```
Comment (mmerce) [inline automl/automl-script/readme.md]: automl-execution

File: automl/automl-script/readme.md [hunk 12]
```diff
@@ -111,23 +125,26 @@ The **outputs** for the script are:
...
 input.  The objective field used in the script will be the one
 associated to the datasets used, so remember to choose them previously
-in both the training and validation datasets.
+in both the training and holdout datasets.
```
Comment (mmerce) [inline automl/automl-script/readme.md]: the training dataset (and holdout dataset when used).

File: automl/automl-script/script.whizzml [hunk 13]
```diff
@@ -171,40 +170,39 @@
...
+                       (if (> total-fields min-fields)
                          (rfe-best-num-features evaluations)
-                         num-fields)))
-    (rfe-output-fields train rfe-output num-features shallow)))
+                         total-fields)))
+    (rfe-output-fields sampled-train rfe-output num-features shallow)))
 
 
 ;; Obtains the most important fields (a list with their names)
 ;; from a train and an optional test dataset
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: remove: and an optional test dataset

File: automl/automl-script/script.whizzml [hunk 14]
```diff
@@ -223,7 +221,7 @@
 ;; Creates the final fusion model with the top 5 models
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: docs says that you use 3 models in the Fusion

File: fuzzy-norms/metadata.json [hunk 15]
```diff
@@ -1,7 +1,8 @@
...
....\n - **[schweizer-sklar p]**: Schweizer–Sklar t-norms. Parameter *p* in\n   the range [-∞, ∞]\n - **[hamacher p]**: Hamacher t-norms. Parameter *p* in the range [0,\n   ∞]\n - **[yager p]**: Yager t-norms. Parameter *p* in the range [0,\n   ∞]\n - **[aczel-alsina p]**: Aczél–Alsina t-norms. Parameter *p* in the range [0,\n   ∞]\n - **[dombi p]**: Dombi t-norms. Parameter *p* in the range [0,\n   ∞]\n - **[sugeno-weber p]**: Sugeno–Weber t-norms. Parameter *p* in the range [-1,\n   ∞]\n\n\nThese are the possible t-conorms:\n\n - **max-tconorm**: Maximum t-norm. Dual to the minimum t-norm, is the\n   smallest t-conorm. No parameters\n - **probabilistic**: Probabilistic t-norm. It's dual to the product\n   t-norm. No parameters.\n - **bounded**: Bounded t-norm. It'ss dual to the Łukasiewicz t-norm. No\n   parameters.\n - **drastic-tconorm**: Drastic t-conorm. It's dual to the drastic\n   t-norm. No paramters.\n - **nilpotent-max**: Nilpotent maximum t-conorm. It's dual to the\n   nilpotent minumum.\n - **einstein-sum**: Einstein t-conorm. It's a dual to one of the Hamacher t-norms. \n\nFind more info in [readme](https://github.com/whizzml/examples/blob/master/fuzzy-norms/readme.md)",
```
Comment (mmerce) [inline fuzzy-norms/metadata.json]: Bellow -> Below

As you can see there are some norms that need a parameter, that...-> As you can see, there are some norms that need a parameter that...

It'ss dual -> It's dual

nilpotent minumum -> nilpotent minimum

**Review by mmerce — CHANGES_REQUESTED** [review]

**Comment by unmonoqueteclea** [comment]
> Let me see if I understand the present status:
> 
>     * We allow the user to set 3 datasets: training, test and holdout
> 
>     * We extend them with the features created from unsupervised learning models
> 
>     * We split the training dataset to hold out a validation-dataset (used only in OptiML selection)
> 
>     * We split the remaining training dataset to hold out a holdout (used only in the Fusion evaluation at the end)
> 
>     * Now we have [training, validation, holdout, (and maybe) test].
> 
> 
> Evaluations:
> 
>     * In RFE we split again the training part to hold out a different validation dataset to do the feature selection.
> 
>     * In OptiML we use the validation dataset held out in step 3 to do the type+parameters selection,
> 
>     * We create a Fusion and an evaluation using the holdout-dataset
>       For this part, every time we run AutoML we get different splits, so non-deterministic behaviour, right? Which I assume is itended to be able to test consistent results over different executions.
> 
> 
> Testing:
> Iff there's a test-dataset:
> 
>     * We create a batch-prediction for the test-dataset based on the previous Fusion
> 
> 
> @unmonoqueteclea, please correct me if I'm wrong

Yes, all of this is mostly right. Only one detail. As you said, we need a holdout dataset to evaluate the final Fusion. This can be given by the user (it's one of the scripts input) but if not, the first thing we do it's obtaining it from the training dataset. The remaining data will be split again to obtain the validation dataset (used in OptiML). And yes, again in the Feature Selection step, the remaining train dataset is split. 

Thank you very much for your exhaustive review, I will check everything tomorrow.

File: automl/automl-library/library.whizzml (hunk 1 above)
Comment (unmonoqueteclea) [reply automl/automl-library/library.whizzml:542]: Yes

**Comment by unmonoqueteclea** [comment]
Ready. Thank you again, @mmerce

**Review by mmerce — APPROVED** [review]
